src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java

```diff
@@ -1059,6 +1083,104 @@ private void handleThresholdResult() {
+            if (response == null) {
+                LOG.warn(NULL_RESPONSE + " ML plugin");
+            } else {
+                pmmlResult = response;
+            }
+            handlePMMLResults();
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            try {
+                handleExecuteException(e, listener, adID);
+            } catch (Exception ex) {
+                LOG.error("Unexpected exception: {} for {}" + ", while calling the ML plugin", ex, adID);
+            } finally {
+                handlePMMLResults();
+            }
+        }
+
+        // Transform result from the ML plugin to the anomaly result format
+        private void handlePMMLResults() {
+            try {
+                if (listener == null) {
+                    LOG.error("Unexpected exception: {} for {}" + ", while calling the ML plugin", "listener is null", adID);
+                    return;
+                }
+                if (pmmlResult == null) {
+                    listener.onFailure(new InternalFailure(adID, "No response from the ML plugin"));
+                    return;
+                }
+                // convert the results from PMML to AD specific format
+                double anomalyGrade = handleGrade(pmmlResult.getOutlier(), pmmlResult.getDecisionFunction());
+                double anomalyScore = handleScore(pmmlResult.getDecisionFunction());
+                // some values are dummy:
+                // 1. Confidence doesn't really change since the ML plugin uploaded models are immutable.
+                double confidence = 1;
+                // 2. PMML models don't have rcfTotalUpdates.
+                long rcfTotalUpdates = 1000;
+                AnomalyResultResponse response = new AnomalyResultResponse(
+                    anomalyGrade,
+                    confidence,
+                    anomalyScore,
+                    features,
+                    null,
+                    rcfTotalUpdates,
+                    detectorIntervalInMinutes
+                );
+                listener.onResponse(response);
+            } catch (Exception e) {
+                handleExecuteException(e, listener, adID);
+            }
+        }
+
+        // Anomaly grade in AD: 0 for normal points, (0, 1] for anomalous points.
+        // Convert [-0.5, 0) to (0, 1] for anomalous points, and [0, 0.5] to 0 for normal points.
+        private double handleGrade(boolean outlier, double decisionFunction) {
+            if (!outlier) {
+                return 0;
+            }
+            return -2 * decisionFunction;
+        }
+
+        // Decision function double from the ML plugin: [-0.5, 0.5]. Negative means anomalous.
+        // Anomaly score in AD is strictly positive. More positive means more anomalous. 1 is a typical cutoff.
+        // Thus, the values can be converted linearly: [-0.5, 0.5] to (0, 2].
+        private double handleScore(double decisionFunction) {
```

Comment:
Anomaly score is not used on frontend, how about we keep the original `decisionFunction` value?

Reply:
During indexing anomaly results, I thought that non-positive values are considered to be erroneous (AnomalyDetectorJobRunner.java -- line 486), so I have to convert this value as well. Could you check and let me know if I misunderstood? Thanks

Reply:
Good callout. You can add this to comment so others know clearly why we have to convert to positive scores

Reply:
Got it!